File: WorldFeed/Server/src/WorldFeed.Common/QueryGenerators/IUserQueryParameterGenerator.cs
﻿namespace WorldFeed.Common.QueryGenerators
{
    using System.Collections.Generic;
    using System.Text;

    using WorldFeed.Common.Public.Models.Interfaces;

    public interface IUserQueryParameterGenerator
    {
        string GenerateIdOrScreenNameParameter(IUserIdentifier user, string idParameterName = "user_id", string screenNameParameterName = "screen_name");
        string GenerateListOfUserIdentifiersParameter(IEnumerable<IUserIdentifier> usersIdentifiers);

        void AppendUser(StringBuilder query, IUserIdentifier user);
        void AppendUsers(StringBuilder query, IEnumerable<IUserIdentifier> users);
    }
}

File: Sharebook/Client/src/app/core/Core/QueryGenerators/IUserQueryParameterGenerator.ts
import {InjectionToken} from "@angular/core";

import {IUserIdentifier} from "../../Public/Models/Interfaces/IUserIdentifier";
import {UserQueryParameterGenerator} from "../../../controllers/User/UserQueryParameterGenerator";
import StringBuilder from "typescript-dotnet-commonjs/System/Text/StringBuilder";

export interface IUserQueryParameterGenerator {
  generateIdOrScreenNameParameter(user: IUserIdentifier, idParameterName?: string, screenNameParameterName?: string): string;

  generateListOfUserIdentifiersParameter(usersIdentifiers: Iterable<IUserIdentifier>): string;

  appendUser(query: StringBuilder, user: IUserIdentifier): void;

  appendUsers(query: StringBuilder, users: Iterable<IUserIdentifier>): void;
}

export const IUserQueryParameterGeneratorToken = new InjectionToken<IUserQueryParameterGenerator>('IUserQueryParameterGenerator', {
  providedIn: 'root',
  factory: () => new UserQueryParameterGenerator(),
});
